### PR TITLE
Fix ray tracing with XR and camera relative rendering

### PR DESCRIPTION
**Pull request by fabien-unity**
---
### Purpose of this PR
Fix XR multi-pass and single-pass modes with ray tracing and camera relative rendering
Fix for https://fogbugz.unity3d.com/f/cases/1336608

---
### Testing status
Tested locally using template scene + ray-traced reflections and AO

---
### Comments to reviewers
I also tested the repro project from the ticket with 2020.3 to confirm it works, but the shaders have changed a bit and it will not be a straight backport.

### CODE REVIEW

File: com.unity.render-pipelines.high-definition/Runtime/RenderPipeline/Raytracing/Shaders/RaytracingAmbientOcclusion.raytrace [hunk 1]
```diff
@@ -89,6 +89,9 @@ void RayGenAmbientOcclusion()
 		rayDescriptor.TMin = 0;
 		rayDescriptor.TMax = _RaytracingRayMaxLength;
 
+        // Adjust world-space position to match the RAS setup with XR single-pass and camera relative
+        ApplyCameraRelativeXR(rayDescriptor.Origin);
```
Comment (anisunity) [inline com.unity.render-pipelines.high-definition/Runtime/RenderPipeline/Raytracing/Shaders/RaytracingAmbientOcclusion.raytrace:93]: small indentation issue

File: com.unity.render-pipelines.high-definition/Runtime/RenderPipeline/Raytracing/Shaders/RaytracingRenderer.raytrace [hunk 2]
```diff
@@ -70,6 +70,9 @@ void RayGenRenderer()
 	rayDescriptor.TMin = _RaytracingCameraNearPlane;
 	rayDescriptor.TMax = _RaytracingRayMaxLength;
 
+    // Adjust world-space position to match the RAS setup with XR single-pass and camera relative
+    ApplyCameraRelativeXR(rayDescriptor.Origin);
```
Comment (anisunity) [inline com.unity.render-pipelines.high-definition/Runtime/RenderPipeline/Raytracing/Shaders/RaytracingRenderer.raytrace:74]: small indentation issue

**Comment by anisunity** [comment]
looks fine to me except for some indentation issues

File: com.unity.render-pipelines.high-definition/Runtime/RenderPipeline/Raytracing/Shaders/RaytracingAmbientOcclusion.raytrace (hunk 1 above)
Comment (fabien-unity) [reply com.unity.render-pipelines.high-definition/Runtime/RenderPipeline/Raytracing/Shaders/RaytracingAmbientOcclusion.raytrace]: fixed

File: com.unity.render-pipelines.high-definition/Runtime/RenderPipeline/Raytracing/Shaders/RaytracingRenderer.raytrace (hunk 2 above)
Comment (fabien-unity) [reply com.unity.render-pipelines.high-definition/Runtime/RenderPipeline/Raytracing/Shaders/RaytracingRenderer.raytrace]: fixed

**Review by sebastienlagarde — APPROVED** [review]

**Comment by m0nsky** [comment]
Hi, sorry to bother. I've tested this locally and can confirm this fixes the issue, but you've forgotten to apply the same changes to RaytracingShadow.raytrace, after that it should be perfect. (Just did it manually and can confirm it will fix the RT shadows in XR)

Thanks for fixing my bugreport so quick, I appreciate it.

**Comment by fabien-unity** [comment]
@m0nsky ah good point! Thanks for reporting and following up. I'll add the code to the shadows tomorrow after some more testing.

**Review by CertifiedRice — APPROVED** [review]

File: com.unity.render-pipelines.high-definition/Runtime/RenderPipeline/Raytracing/Shaders/Deferred/RaytracingGBuffer.raytrace [hunk 3]
```diff
@@ -116,6 +116,9 @@ void TraceGBuffer(PixelCoordinates coords)
     rayDescriptor.Direction = direction.xyz;
     rayDescriptor.TMin = 0.0;
     rayDescriptor.TMax = _RaytracingRayMaxLength;
+    
+    // Adjust world-space position to match the RAS setup with XR single-pass and camera relative
+    ApplyCameraRelativeXR(rayDescriptor.Origin);
```
Comment (sebastienlagarde) [inline com.unity.render-pipelines.high-definition/Runtime/RenderPipeline/Raytracing/Shaders/Deferred/RaytracingGBuffer.raytrace:121]: @anisunity / @fabien-unity  by doing call like this we will always make mistake with copy/paste 

As far as I can see the code is always the same. We should do a function and call it.

RayDesc rayDescriptor;
InitRayDesc(rayDescriptor, posInput.positionWS, normalData.normalWS, direction);

with 

InitRayDesc(RayDesc  rayDescriptor, ...)
{
    rayDescriptor.Origin = ApplyCameraRelativeXR(posInput.positionWS + normalData.normalWS * _RaytracingRayBias);
    rayDescriptor.Direction = direction.xyz;
    rayDescriptor.TMin = 0.0;
    rayDescriptor.TMax = _RaytracingRayMaxLength;
}

Thought?

Comment (fabien-unity) [reply com.unity.render-pipelines.high-definition/Runtime/RenderPipeline/Raytracing/Shaders/Deferred/RaytracingGBuffer.raytrace:121]: Yes that'll be ideal. I can refactor the code, unless @anisunity wants to do it :)

Comment (anisunity) [reply com.unity.render-pipelines.high-definition/Runtime/RenderPipeline/Raytracing/Shaders/Deferred/RaytracingGBuffer.raytrace:121]: i'll Do it. Does it have to be in this PR?

Comment (anisunity) [reply com.unity.render-pipelines.high-definition/Runtime/RenderPipeline/Raytracing/Shaders/Deferred/RaytracingGBuffer.raytrace:121]: @fabien-unity

Comment (fabien-unity) [reply com.unity.render-pipelines.high-definition/Runtime/RenderPipeline/Raytracing/Shaders/Deferred/RaytracingGBuffer.raytrace:121]: I'm ok with the refactor being done after this PR lands.

Comment (sebastienlagarde) [reply com.unity.render-pipelines.high-definition/Runtime/RenderPipeline/Raytracing/Shaders/Deferred/RaytracingGBuffer.raytrace:121]: alright, landing this PR then :)

**Pull request merged**